core/src/replay_stage.rs

```diff
@@ -864,53 +992,211 @@ impl ReplayStage {
+                        "{} slot_weight: {} {} {} {}",
+                        my_pubkey,
+                        stats.slot,
+                        stats.weight,
+                        stats.fork_weight,
+                        bank.parent().map(|b| b.slot()).unwrap_or(0)
+                    );
+                    stats.stake_lockouts = stake_lockouts;
+                    stats.block_height = bank.block_height();
+                    stats.computed = true;
+                    new_stats.push(stats.slot);
+                }
+            }
+
+            Self::update_propagation_status(
+                progress,
+                bank_slot,
+                all_pubkeys,
+                bank_forks,
+                vote_tracker,
+            );
+
             let stats = &mut progress
-                .get_mut(&bank.slot())
+                .get_mut(&bank_slot)
                 .expect("All frozen banks must exist in the Progress map")
                 .fork_stats;
-
-            if !stats.computed {
-                stats.slot = bank.slot();
-                let (stake_lockouts, total_staked, bank_weight) = tower.collect_vote_lockouts(
-                    bank.slot(),
-                    bank.vote_accounts().into_iter(),
-                    &ancestors,
-                );
-                stats.total_staked = total_staked;
-                stats.weight = bank_weight;
-                stats.fork_weight = stats.weight + parent_weight;
-
-                datapoint_info!(
-                    "bank_weight",
-                    ("slot", bank.slot(), i64),
-                    // u128 too large for influx, convert to hex
-                    ("weight", format!("{:X}", stats.weight), String),
-                );
-                info!(
-                    "{} slot_weight: {} {} {} {}",
-                    my_pubkey,
-                    stats.slot,
-                    stats.weight,
-                    stats.fork_weight,
-                    bank.parent().map(|b| b.slot()).unwrap_or(0)
-                );
-                stats.stake_lockouts = stake_lockouts;
-                stats.block_height = bank.block_height();
-                stats.computed = true;
-                new_stats.push(stats.slot);
-            }
             stats.vote_threshold = tower.check_vote_stake_threshold(
-                bank.slot(),
+                bank_slot,
                 &stats.stake_lockouts,
                 stats.total_staked,
             );
-            stats.is_locked_out = tower.is_locked_out(bank.slot(), &ancestors);
-            stats.has_voted = tower.has_voted(bank.slot());
-            stats.is_recent = tower.is_recent(bank.slot());
+            stats.is_locked_out = tower.is_locked_out(bank_slot, &ancestors);
+            stats.has_voted = tower.has_voted(bank_slot);
+            stats.is_recent = tower.is_recent(bank_slot);
         }
         new_stats
     }
 
+    fn is_propagated(progress: &mut ProgressMap, slot: Slot) -> bool {
```

Comment:
Can you pull these big functions into their own object in a separate file?  Kind of like how i did with ClusterSlots